src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,520 @@
+        {
+            // TP optimization - skip blocks that *likely* don't have bounds checks
+            continue;
+        }
+
+        if (block->isRunRarely())
+        {
+            continue;
+        }
+
+        if (!block->KindIs(BBJ_ALWAYS, BBJ_RETURN, BBJ_COND))
+        {
+            // For now, we support only simple blocks
+            continue;
+        }
+
+        bndChkLocations.Reset();
+        bndChkMap.RemoveAll();
+
+        for (Statement* const stmt : block->Statements())
+        {
+            if (block->KindIs(BBJ_COND, BBJ_RETURN) && (stmt == block->lastStmt()))
+            {
+                // TODO-RangeCheckCloning: Splitting these blocks at the last statements
+                // require using gtSplitTree for the last bounds check.
+                break;
+            }
+
+            // Now just record all the bounds checks in the block (in the execution order)
+            //
+            BoundsChecksVisitor visitor(this, stmt, &bndChkLocations);
+            visitor.WalkTree(stmt->GetRootNodePointer(), nullptr);
+        }
+
+        if (bndChkLocations.Height() < MIN_CHECKS_PER_GROUP)
+        {
+            JITDUMP("Not enough bounds checks in the block - bail out.\n");
+            continue;
+        }
+
+        // Now we need to group the bounds checks by the base index and length VNs.
+        // We could do it directly in the visitor above and avoid this O(n) pass,
+        // but it's more TP/Memory wise to use stack-allocated ArrayStack first and
+        // bail out on <MIN_CHECKS_PER_GROUP cases.
+        for (int i = 0; i < bndChkLocations.Height(); i++)
+        {
+            BoundCheckLocation loc = bndChkLocations.Bottom(i);
+            BoundsCheckInfo    bci{};
+            if (bci.Initialize(this, loc.stmt, loc.bndChk, loc.bndChkParent))
+            {
+                IdxLenPair            key(bci.idxVN, bci.lenVN);
+                BoundsCheckInfoStack* value;
+                if (!bndChkMap.Lookup(key, &value))
+                {
+                    CompAllocator allocator = getAllocator(CMK_RangeCheckCloning);
+                    value                   = new (allocator) BoundsCheckInfoStack(allocator);
+                    bndChkMap.Set(key, value);
+                }
+                value->Push(bci);
+            }
+        }
+
+        if (bndChkMap.GetCount() == 0)
+        {
+            JITDUMP("No bounds checks in the block - bail out.\n");
+            continue;
+        }
+
+        // Now choose the largest group of bounds checks (the one with the most checks)
+        BoundsCheckInfoStack* largestGroup = nullptr;
+        for (BoundsCheckInfoMap::Node* keyValuePair : BoundsCheckInfoMap::KeyValueIteration(&bndChkMap))
+        {
+            ArrayStack<BoundsCheckInfo>* value = keyValuePair->GetValue();
+            if ((largestGroup == nullptr) || (value->Height() > largestGroup->Height()))
+            {
+                largestGroup = value;
+            }
+        }
```

Comment:
Why not do it for all groups? It seems arbitrary that if I split a block then the optimization might kick in more.

Reply:
I'd like to try that in a separate PR to estimate an impact if you don't mind - one of my intermediate commit supported it

Reply:
One problem is how to propagate side effects from the first bounds checks of secondary groups. Today, I just do `gtSplitTree` on the first bounds check and it handles side effect ordering issue for me naturally